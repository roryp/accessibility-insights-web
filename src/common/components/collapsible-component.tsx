// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@uifabric/utilities';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        const newState = !this.state.showContent;
        this.setState({ showContent: newState });
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, 'collapsible-content')}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName, 'collapsible-component')}>
                <ActionButton
                    className="collapsible"
                    iconProps={{ iconName: iconName, class: 'collapsible-icon' }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    <span className="collapsible-title">{this.props.header}</span>
                </ActionButton>
                {content}
            </div>
        );
    }
}
