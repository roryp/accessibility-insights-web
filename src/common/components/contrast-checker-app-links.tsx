// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { windowsProductName } from 'content/strings/application';
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { NewTabLink } from './new-tab-link';

export const WindowsContrastCheckerAppLink = NamedFC('WindowsContrastCheckerAppLink', () => (
    <NewTabLink href="https://go.microsoft.com/fwlink/?linkid=2075365">
        {windowsProductName}
    </NewTabLink>
));

export const MacContrastCheckerAppLink = NamedFC('MacContrastCheckerAppLink', () => (
    <NewTabLink href="https://developer.paciellogroup.com/resources/contrastanalyser/">
        Colour Contrast Analyser
    </NewTabLink>
));
