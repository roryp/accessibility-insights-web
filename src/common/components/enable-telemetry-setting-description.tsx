// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { PrivacyStatementText, PrivacyStatementTextDeps } from './privacy-statement-text';
import { TelemetryNotice, TelemetryNoticeDeps } from './telemetry-notice';

export type EnableTelemetrySettingDescriptionDeps = TelemetryNoticeDeps & PrivacyStatementTextDeps;

export type EnableTelemetrySettingDescriptionProps = {
    deps: EnableTelemetrySettingDescriptionDeps;
};

export const EnableTelemetrySettingDescription = NamedFC<EnableTelemetrySettingDescriptionProps>(
    'EnableTelemetrySettingDescription',
    props => (
        <>
            <TelemetryNotice deps={props.deps} />
            <PrivacyStatementText deps={props.deps} />
        </>
    ),
);
