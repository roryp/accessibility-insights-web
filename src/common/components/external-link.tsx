// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link } from 'office-ui-fabric-react';
import * as React from 'react';

import { ActionInitiators } from '../action/action-initiator';
import { NamedFC } from '../react/named-fc';

export type ExternalLinkDeps = {
    actionInitiators: Pick<ActionInitiators, 'openExternalLink'>;
};

export type ExternalLinkProps = {
    deps: ExternalLinkDeps;
    href: string;
    title?: string;
};

export const ExternalLink = NamedFC<ExternalLinkProps>(
    'ExternalLink',
    ({ deps, href, title, children }) => {
        const onClick = e => deps.actionInitiators.openExternalLink(e, { href });
        return (
            <Link
                className="insights-link"
                target="_blank"
                href={href}
                title={title}
                onClick={onClick}
            >
                {children}
            </Link>
        );
    },
);
