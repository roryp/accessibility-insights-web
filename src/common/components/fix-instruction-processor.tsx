// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import * as styles from './cards/fix-instruction-color-box.scss';

type ColorMatch = {
    splitIndex: number;
    colorHexValue: string;
};

export class FixInstructionProcessor {
    private readonly colorValueMatcher = `(#[0-9a-f]{6})`;
    private readonly foregroundColorText = 'foreground color: ';
    private readonly foregroundRegExp = new RegExp(
        `${this.foregroundColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundColorText = 'background color: ';
    private readonly backgroundRegExp = new RegExp(
        `${this.backgroundColorText}${this.colorValueMatcher}`,
        'i',
    );

    public process(fixInstruction: string): JSX.Element {
        const matches = this.getColorMatches(fixInstruction);
        return this.splitFixInstruction(fixInstruction, matches);
    }

    private getColorMatches(fixInstruction: string): ColorMatch[] {
        const foregroundMatch = this.getColorMatch(fixInstruction, this.foregroundRegExp);
        const backgroundMatch = this.getColorMatch(fixInstruction, this.backgroundRegExp);

        return [foregroundMatch, backgroundMatch].filter(match => match != null) as ColorMatch[];
    }

    private getColorMatch(fixInstruction: string, colorRegex: RegExp): ColorMatch | null {
        if (!colorRegex.test(fixInstruction)) {
            return null;
        }

        const match = colorRegex.exec(fixInstruction);
        if (match == null || match[1] == null) {
            return null;
        }

        const colorHexValue = match[1];

        return {
            splitIndex: match.index + this.foregroundColorText.length + colorHexValue.length,
            colorHexValue,
        };
    }

    private splitFixInstruction(fixInstruction: string, matches: ColorMatch[]): JSX.Element {
        const sortedMatches = matches.sort((a, b) => a.splitIndex - b.splitIndex);

        if (sortedMatches.length === 0) {
            return <>{fixInstruction}</>;
        }

        let insertionIndex = 0;
        let keyIndex = 0;

        const result: JSX.Element[] = [];

        sortedMatches.forEach(match => {
            const endIndex = match.splitIndex - match.colorHexValue.length;
            const substring = fixInstruction.substring(insertionIndex, endIndex);

            result.push(<span key={`instruction-split-${keyIndex++}`}>{substring}</span>);

            result.push(this.createColorBox(match.colorHexValue, keyIndex++));

            insertionIndex = endIndex;
        });

        const coda = fixInstruction.substr(insertionIndex);

        result.push(<span key={`instruction-split-${keyIndex++}`}>{coda}</span>);

        return (
            <>
                <span aria-hidden="true">{result}</span>
                <span className={styles.screenReaderOnly}>{fixInstruction}</span>
            </>
        );
    }

    private createColorBox(colorHexValue: string, keyIndex: number): JSX.Element {
        return (
            <span
                key={`instruction-split-${keyIndex}`}
                className={styles.fixInstructionColorBox}
                style={{ backgroundColor: colorHexValue }}
            />
        );
    }
}
