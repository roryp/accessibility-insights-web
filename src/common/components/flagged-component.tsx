// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { FeatureFlagStoreData } from '../types/store-data/feature-flag-store-data';

export interface FlaggedComponentProps {
    featureFlagStoreData: FeatureFlagStoreData;
    featureFlag: string;
    enableJSXElement: JSX.Element;
    disableJSXElement?: JSX.Element;
}

export class FlaggedComponent extends React.Component<FlaggedComponentProps> {
    public render(): JSX.Element | null {
        const flagName = this.props.featureFlag;

        if (this.props.featureFlagStoreData != null && this.props.featureFlagStoreData[flagName]) {
            return this.props.enableJSXElement;
        }

        return this.props.disableJSXElement || null;
    }
}
