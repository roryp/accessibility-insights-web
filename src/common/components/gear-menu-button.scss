// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../common/styles/colors.scss';

:global(.ms-Button).gear-menu-button:focus::after {
    border: 1px $neutral-0 solid;
}

:global(.high-contrast-theme) {
    .gear-menu-button {
        color: $neutral-alpha-90;
    }

    .gear-menu-button-callout {
        box-shadow: $always-white 0px 0px 1px 0px !important;
        border-radius: unset;
    }
}
