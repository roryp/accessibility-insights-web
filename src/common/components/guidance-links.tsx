// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { LinkComponentType } from 'common/types/link-component-type';
import { isEmpty } from 'lodash';
import * as React from 'react';
import { HyperlinkDefinition } from 'views/content/content-page';

import { NamedFC } from '../react/named-fc';

export interface GuidanceLinksProps {
    links: HyperlinkDefinition[];
    classNameForDiv?: string;
    LinkComponent: LinkComponentType;
}

export const GuidanceLinks = NamedFC('GuidanceLinks', (props: GuidanceLinksProps) => {
    const { links, classNameForDiv } = props;

    if (isEmpty(links)) {
        return null;
    }

    const renderLinks = (): JSX.Element[] => {
        return links.map((link, index) => {
            return renderLink(link, index, links.length);
        });
    };

    const renderLink = (link: HyperlinkDefinition, index: number, length: number): JSX.Element => {
        const addComma: boolean = index !== length - 1;
        const comma = addComma ? <span>,&nbsp;</span> : null;
        const LinkComponent: LinkComponentType = props.LinkComponent;
        return (
            <React.Fragment key={`guidance-link-${index}`}>
                <LinkComponent href={link.href} onClick={event => event.stopPropagation()}>
                    {link.text.toUpperCase()}
                </LinkComponent>
                {comma}
            </React.Fragment>
        );
    };

    const spanClassName = classNameForDiv || 'guidance-links';
    return <span className={spanClassName}>{renderLinks()}</span>;
});
