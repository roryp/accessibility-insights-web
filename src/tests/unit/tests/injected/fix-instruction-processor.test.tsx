// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FixInstructionProcessor } from 'common/components/fix-instruction-processor';

describe('FixInstructionProcessor', () => {
    let testSubject: FixInstructionProcessor;

    beforeEach(() => {
        testSubject = new FixInstructionProcessor();
    });

    test('no background nor foreground on the message', () => {
        const fixInstruction =
            'there is nothing that will trigger the processor to change the fix instruction here';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('foreground color on the message', () => {
        const fixInstruction =
            'color contrast of 2.52 (foreground color: #8a94a8; nothing else to match here)';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('background color on the message', () => {
        const fixInstruction =
            'color contrast of 2.52 (background color: #8a94a8; nothing else to match here)';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('foreground and background color on the message (mind the order)', () => {
        const fixInstruction =
            'Element has insufficient color contrast of 2.52 (foreground color: #8a94a8, background color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('background and foreground on the message (mind the order)', () => {
        const fixInstruction =
            'Element has insufficient color contrast of 2.52 (background color: #8a94a8, foreground color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('we assume only one instance of fore/background color, second instance will be ignored', () => {
        const fixInstruction =
            'first foreground color: #ffffff, second foreground: #000000, first background color: #AAAAAA, second background color: #bbBBbb';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });

    test('when we have nothing to process', () => {
        const fixInstruction = 'nothing to process';

        const result = testSubject.process(fixInstruction);

        expect(result).toMatchSnapshot();
    });
});
