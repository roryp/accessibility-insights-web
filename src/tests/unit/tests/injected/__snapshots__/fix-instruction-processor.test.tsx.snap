// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixInstructionProcessor background and foreground on the message (mind the order) 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      Element has insufficient color contrast of 2.52 (background color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#8a94a8",
        }
      }
    />
    <span>
      #8a94a8, foreground color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#e7ecd8",
        }
      }
    />
    <span>
      #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    Element has insufficient color contrast of 2.52 (background color: #8a94a8, foreground color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor background color on the message 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      color contrast of 2.52 (background color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#8a94a8",
        }
      }
    />
    <span>
      #8a94a8; nothing else to match here)
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    color contrast of 2.52 (background color: #8a94a8; nothing else to match here)
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor foreground and background color on the message (mind the order) 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      Element has insufficient color contrast of 2.52 (foreground color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#8a94a8",
        }
      }
    />
    <span>
      #8a94a8, background color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#e7ecd8",
        }
      }
    />
    <span>
      #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    Element has insufficient color contrast of 2.52 (foreground color: #8a94a8, background color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor foreground color on the message 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      color contrast of 2.52 (foreground color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#8a94a8",
        }
      }
    />
    <span>
      #8a94a8; nothing else to match here)
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    color contrast of 2.52 (foreground color: #8a94a8; nothing else to match here)
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor no background nor foreground on the message 1`] = `
<React.Fragment>
  there is nothing that will trigger the processor to change the fix instruction here
</React.Fragment>
`;

exports[`FixInstructionProcessor we assume only one instance of fore/background color, second instance will be ignored 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      first foreground color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    />
    <span>
      #ffffff, second foreground: #000000, first background color: 
    </span>
    <span
      className="fixInstructionColorBox"
      style={
        Object {
          "backgroundColor": "#AAAAAA",
        }
      }
    />
    <span>
      #AAAAAA, second background color: #bbBBbb
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    first foreground color: #ffffff, second foreground: #000000, first background color: #AAAAAA, second background color: #bbBBbb
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor when we have nothing to process 1`] = `
<React.Fragment>
  nothing to process
</React.Fragment>
`;
